pkg/monitortestlibrary/pathologicaleventlibrary/duplicated_events.go

```diff
@@ -252,9 +260,68 @@ func (d duplicateEventsEvaluator) testDuplicatedEvents(testName string, flakeOnl
+			// Build the allowed time range only once
+			if buildTopologyHintAllowedTimeRange {
+				taintManagerTestIntervals := events.Filter(func(eventInterval monitorapi.Interval) bool {
+					return eventInterval.Source == monitorapi.SourceE2ETest &&
+						strings.Contains(eventInterval.StructuredLocator.Keys[monitorapi.LocatorE2ETestKey], "NoExecuteTaintManager")
+				})
+				// Start the allowed time range from time range of the tests. But events lag behind the tests since the tests do not wait
+				// until all dns pods are properly scheduled and reach ready state. So we will need to expand the allowed time range after.
+				for _, test := range taintManagerTestIntervals {
+					topologyHintAllowedTimeRange = append(topologyHintAllowedTimeRange, &timeRange{from: test.From, to: test.To})
+				}
+				dnsUpdateIntervals := events.Filter(func(eventInterval monitorapi.Interval) bool {
+					return eventInterval.Source == monitorapi.SourcePodState &&
+						(eventInterval.StructuredLocator.Type == monitorapi.LocatorTypePod || eventInterval.StructuredLocator.Type == monitorapi.LocatorTypeContainer) &&
+						eventInterval.StructuredLocator.Keys[monitorapi.LocatorNamespaceKey] == "openshift-dns" &&
+						eventInterval.StructuredMessage.Annotations[monitorapi.AnnotationConstructed] == monitorapi.ConstructionOwnerPodLifecycle
+				})
+
+				// Now expand the allowed time range until the replacement dns pod gets ready
+				for _, r := range topologyHintAllowedTimeRange {
+					var lastReadyTime time.Time
+					count := 0
+					for _, interval := range dnsUpdateIntervals {
+						if interval.From.Before(r.from) {
+							continue
+						}
+						// If there is a GracefulDelete of dns-default pod, we will have to wait until the replacement dns container becomes ready
+						if interval.StructuredMessage.Reason == monitorapi.PodReasonGracefulDeleteStarted &&
+							strings.Contains(interval.StructuredLocator.Keys[monitorapi.LocatorPodKey], "dns-default") {
+							count++
+						}
+						if interval.StructuredMessage.Reason == monitorapi.ContainerReasonReady &&
+							interval.StructuredLocator.Keys[monitorapi.LocatorContainerKey] == "dns" && count > 0 {
+							lastReadyTime = interval.From
+							count--
+						}
+						if interval.From.After(r.to) && count == 0 {
+							if lastReadyTime.After(r.to) {
+								r.to = lastReadyTime
+							}
+							break
+						}
+					}
+				}
+				buildTopologyHintAllowedTimeRange = false
+			}
+			// Filter out TopologyAwareHintsDisabled events within allowed time range
+			var allowed bool
+			for _, r := range topologyHintAllowedTimeRange {
+				if r.from.Before(event.From) && r.to.After(event.To) {
+					logrus.Infof("%s was found to fall into the allowed time range %+v, ignoring pathological event as we expect these during NoExecuteTaintManager test", event, r)
+					allowed = true
+					break
+				}
+			}
+			if allowed {
+				continue
+			}
+		}
```

Comment:
I'm working on pathological events quite heavily but the way you've done this will port over nicely.